****
**Instituto Federal de Educação, Ciência e Tecnologia de Minas Gerais - Campus Ouro Preto**<br>
**Especialização em Inteligência Artificial**<br>
**Disciplina: Redes Neurais e Aprendizado Profundo**<br>
**Profs.: Dr. Agnaldo Silva, Dr. Frederico Gadelha, Dra. Sílvia Almeida**<br>
**Alunos:  Fernando Fernandes, Ivanete e Marco Antônio**
****

1. O que é inteligência para você(s)?

*O conceito de inteligência está associado à capacidade humana de tomar decisões e resolver problemas; à capacidade de se adaptar a diferentes situações; à capacidade de aprender algo novo, a partir da detecção de padrões;*

2. Em sua opinião (ou na do grupo), o que aconteceria se alguém descobrisse como implementar uma IA mais abrangente (e.g., AGI) em um robô?

*Em algumas tarefas, as máquinas já são capazes de ter um desempenho semelhante ou melhor do que o ser humano, por exemplo, no processamento e análise de dados e imagens. De qualquer forma, os modelos de inteligência artificial atuais são capazes de resolver tarefas específicas. No momento em que for possível o desenvolvimento e implantação de uma IA mais abrangente em um robô, principalmente, se (ou quando) alcançarmos uma AGI (Artificial General Intelligence), para alguns pesquisadores, estaremos diante de um risco à raça humana; para outros, esse é o propósito das pesquisas em Inteligência Artificial.*

3. A partir da análise de um processo de destilação fracionada de petróleo observou-se que determinado óleo poderia ser classificado em duas classes de pureza {C1 e C2}, mediante a medição de três grandezas {x1, x2
e x3} que representam algumas das propriedades físico-químicas do óleo. Para tanto, pretende-se utilizar um perceptron para executar a classificação automática dessas duas classes. Assim, baseadas nas informações
coletadas do processo, formou-se o conjunto de treinamento em anexo1, tomando por convenção o valor –1 para óleo pertencente à classe C1 e o valor +1 para óleo pertencente à classe C2.

Daí, pede-se:

a. Execute dois treinamentos para a rede perceptron, inicializando-se o vetor de pesos em cada treinamento com valores aleatórios entre zero e um de tal forma que os elementos do vetor de pesos iniciais não sejam os mesmos.

In [1]:
import numpy as np

class Perceptron:
    def __init__(self, num_features, learning_rate=0.01, epochs=100):
        self.num_features = num_features
        self.learning_rate = learning_rate
        self.epochs = epochs
        self.weights = np.random.rand(num_features + 1)  # initial random weights +1 for the bias term
        self.initialWeights = self.weights
        print(f'[INFO] \tRandom initial weights: {self.weights}')

    def predict(self, inputs): # activation function
        summation = np.dot(inputs, self.weights[1:]) + self.weights[0] # activation potential: u
        return 1 if summation >= 0 else -1 # Use of the bipolar step function

    def train(self, training_data, labels):
        hasError = True
        for epoch in range(self.epochs):
            #print(f'[INFO] Epoch: {epoch}')
            hasError = False
            for inputs, label in zip(training_data, labels):
                prediction = self.predict(inputs) # return of activation function: y
                if prediction != label:
                    hasError = True
                    update = self.learning_rate * (label - prediction) # eta * (dk - y)
                    self.weights[1:] += update * inputs # update weights: w <- w + eta * (dk - y) * xk
                    self.weights[0] += update # update activation limiar: tetha <- tetha + eta * (dk - y)
                    #print(f'[INFO] Weights: {self.weights}')
            if hasError == False:
                print(f'[INFO] \tConverged after: {epoch + 1} epochs.')
                break
        print(f'[INFO] \tFinal weights: {self.weights}')
        print(f'[INFO] \tTotal of epochs: {epoch + 1}')
    
    def getInitialWeights(self):
        return self.initialWeights

    def getFinalWeights(self):
        return self.weights

In [2]:
print(f'\n[INFO] ###### Perceptron Implementation #######')
print(f'\n[INFO] Loading training dataset and labels...')
file = open('tab_treinamento1.dat', 'r')
results = list()
l = list()

for line in file:
    columns = line.split()
    columns = np.array(columns, dtype=float)
    results.append(columns[:3])
    l.append(columns[-1:])
    
training_data = np.array(results)
labels = np.array(l)
print(f'\t[INFO] OK!')

#training_data = np.array([[0.6508, 0.1097, 4.0009], [-1.4492, 0.8896, 4.4005], [2.085, 0.6876, 1.2071], [0.2626, 1.1476, 7.7985], [0.6418, 1.0234, 7.0427], [0.2569, 0.673, 8.3265], [1.1155, 0.6043, 7.4446], [0.0914, 0.3399, 7.0677], [0.0121, 0.5256, 4.6316], [-0.0429, 0.466, 5.4323], [0.434, 0.687, 8.2287], [0.2735, 1.0287, 7.1934], [0.4839, 0.4851, 7.485], [0.4089, -0.1267, 5.5019], [1.4391, 0.1614, 8.5843], [-0.9115,  -0.1973, 2.1962], [0.3654, 1.0475, 7.4858], [0.2144, 0.7515, 7.1699], [0.2013, 1.0014, 6.5489], [0.6483, 0.2183, 5.8991], [-0.1147, 0.2242, 7.2435], [-0.797, 0.8795, 3.8762], [-1.0625, 0.6366, 2.4707], [0.5307, 0.1285, 5.6883], [-1.22, 0.7777, 1.7252], [0.3957, 0.1076, 5.6623], [-0.1013, 0.5989, 7.1812], [2.4482, 0.9455, 11.2095], [2.0149, 0.6192, 10.9263], [0.2012, 0.2611, 5.4631]])
#labels = np.array([-1, -1, -1, 1, 1, -1, 1, -1, 1, 1, -1, 1, -1, -1, -1, -1, 1, 1, 1, 1, -1, 1, 1, 1, 1, -1, -1, 1, -1, 1])
print(f'\n[INFO] Getting information about training dataset...') 
print(f'[INFO] Training dataset: \n{training_data}')
print(f'[INFO] Labels of training dataset: \n{labels}')


[INFO] ###### Perceptron Implementation #######

[INFO] Loading training dataset and labels...
	[INFO] OK!

[INFO] Getting information about training dataset...
[INFO] Training dataset: 
[[-0.6508  0.1097  4.0009]
 [-1.4492  0.8896  4.4005]
 [ 2.085   0.6876 12.071 ]
 [ 0.2626  1.1476  7.7985]
 [ 0.6418  1.0234  7.0427]
 [ 0.2569  0.673   8.3265]
 [ 1.1155  0.6043  7.4446]
 [ 0.0914  0.3399  7.0677]
 [ 0.0121  0.5256  4.6316]
 [-0.0429  0.466   5.4323]
 [ 0.434   0.687   8.2287]
 [ 0.2735  1.0287  7.1934]
 [ 0.4839  0.4851  7.485 ]
 [ 0.4089 -0.1267  5.5019]
 [ 1.4391  0.1614  8.5843]
 [-0.9115 -0.1973  2.1962]
 [ 0.3654  1.0475  7.4858]
 [ 0.2144  0.7515  7.1699]
 [ 0.2013  1.0014  6.5489]
 [ 0.6483  0.2183  5.8991]
 [-0.1147  0.2242  7.2435]
 [-0.797   0.8795  3.8762]
 [-1.0625  0.6366  2.4707]
 [ 0.5307  0.1285  5.6883]
 [-1.22    0.7777  1.7252]
 [ 0.3957  0.1076  5.6623]
 [-0.1013  0.5989  7.1812]
 [ 2.4482  0.9455 11.2095]
 [ 2.0149  0.6192 10.9263]
 [ 0.2012  0.2611  5.4631]]
[

In [94]:
# Creating a Perceptron
print(f'\n[INFO] Creating a Perceptron...')
number_of_epochs = 10000
perceptron1 = Perceptron(num_features=3, learning_rate=0.01, epochs=number_of_epochs)
print(f'[INFO] \tOK!')


[INFO] Creating a Perceptron...
[INFO] 	Random initial weights: [0.17736342 0.15638614 0.47744044 0.1209934 ]
[INFO] 	OK!


In [95]:
# Training the perceptron 1
print(f'\n[INFO] Getting information about training dataset...')
print(f'[INFO] \tTraining dataset size = {training_data.shape[0]}')
print(f'[INFO] \tLabels size = {labels.shape[0]}')
print(f'[INFO] \tLimit of epochs: {number_of_epochs}')
print(f'\n[INFO] Training the Perceptron 1...')
perceptron1.train(training_data, labels)

initialWeights = perceptron1.getInitialWeights()
finalWeights = perceptron1.getFinalWeights()
print(f'\t[INFO] OK!')


[INFO] Getting information about training dataset...
[INFO] 	Training dataset size = 30
[INFO] 	Labels size = 30
[INFO] 	Limit of epochs: 10000

[INFO] Training the Perceptron 1...
[INFO] 	Converged after: 333 epochs.
[INFO] 	Final weights: [ 2.93736342  1.41135414  2.43657844 -0.7023866 ]
[INFO] 	Total of epochs: 333
	[INFO] OK!


In [96]:
# Creating a new Perceptron
print(f'\n[INFO] Creating a new Perceptron...')
number_of_epochs = 10000
perceptron2 = Perceptron(num_features=3, learning_rate=0.01, epochs=number_of_epochs)
print(f'[INFO] \tOK!')


[INFO] Creating a new Perceptron...
[INFO] 	Random initial weights: [0.44467875 0.60329962 0.69505087 0.33930767]
[INFO] 	OK!


In [97]:
# Training the perceptron 2
print(f'\n[INFO] Getting information about training dataset...')
print(f'[INFO] \tTraining dataset size = {training_data.shape[0]}')
print(f'[INFO] \tLabels size = {labels.shape[0]}')
print(f'[INFO] \tLimit of epochs: {number_of_epochs}')
print(f'\n[INFO] Training the Perceptron 2...')
perceptron2.train(training_data, labels)

print(f'\t[INFO] OK!')


[INFO] Getting information about training dataset...
[INFO] 	Training dataset size = 30
[INFO] 	Labels size = 30
[INFO] 	Limit of epochs: 10000

[INFO] Training the Perceptron 2...
[INFO] 	Converged after: 386 epochs.
[INFO] 	Final weights: [ 3.06467875  1.55732162  2.47131887 -0.73088233]
[INFO] 	Total of epochs: 386
	[INFO] OK!


b. Registre os resultados dos dois treinamentos na tabela a seguir:

Tabela 1 - Resultados dos treinamentos (**tab_treinamento1.dat**):
<table border="1">
	<tr>
		<td rowspan="2">Treinamento</td>
		<td colspan="4">Vetor de Pesos Inicial</td>
		<td colspan="4">Vetor de Pesos Final</td>
		<td rowspan="2"> Número de Épocas</td>
	</tr>
	<tr>
		<td>b</td>
		<td>w1</td>
		<td>w2</td>
		<td>w3</td>
		<td>b</td>
		<td>w1</td>
		<td>w2</td>
		<td>w3</td>
	</tr>
	<tr>
		<td>1º (T1)</td>
		<td>0.1773634</td>
		<td>0.15638614</td>
		<td>0.47744044</td>
		<td>0.1209934</td>
		<td>2.93736342</td>
		<td>1.41135414</td>
		<td>2.43657844</td>
		<td>-0.7023866</td>
		<td>333</td>
	</tr>
	<tr>
		<td>2º (T1)</td>
		<td>0.44467875</td>
		<td>0.15638614</td>
		<td>0.69505087</td>
		<td>0.33930767</td>
		<td>3.06467875</td>
		<td>1.55732162</td>
		<td>2.47131887</td>
		<td>-0.73088233</td>
		<td>386</td>
	</tr>
</table>

*Nota: os valores atuais dessa tabela se referem ao últimos treinamentos realizados. Caso os trechos de código acima sejam executados novamente, necessita-se atualizar a tabela.*

c. Após o treinamento do perceptron, aplique-o na classificação automática de novas amostras de óleo (ver arquivo tab_teste1.dat), indicando-se na tabela seguinte os resultados das saídas (Classes) referentes aos dois processos de treinamento realizados no item a.

In [98]:
# Testing dataset
# test_data = np.array([[0.6508, 0.1097, 4.0009], [-1.4492, 0.8896, 4.4005], [2.085, 0.6876, 1.2071], [0.2626, 1.1476, 7.7985], [0.6418, 1.0234, 7.0427], [0.2569, 0.673, 8.3265], [1.1155, 0.6043, 7.4446], [0.0914, 0.3399, 7.0677], [0.0121, 0.5256, 4.6316], [-0.0429, 0.466, 5.4323], [0.434, 0.687, 8.2287], [0.2735, 1.0287, 7.1934], [0.4839, 0.4851, 7.485], [0.4089, -0.1267, 5.5019], [1.4391, 0.1614, 8.5843], [-0.9115,  -0.1973, 2.1962], [0.3654, 1.0475, 7.4858], [0.2144, 0.7515, 7.1699], [0.2013, 1.0014, 6.5489], [0.6483, 0.2183, 5.8991], [-0.1147, 0.2242, 7.2435], [-0.797, 0.8795, 3.8762], [-1.0625, 0.6366, 2.4707], [0.5307, 0.1285, 5.6883], [-1.22, 0.7777, 1.7252], [0.3957, 0.1076, 5.6623], [-0.1013, 0.5989, 7.1812], [2.4482, 0.9455, 11.2095], [2.0149, 0.6192, 10.9263], [0.2012, 0.2611, 5.4631]])

# Testing the perceptron
print(f'\n[INFO] Loading testing dataset...')
file = open('tab_teste1.dat', 'r')
results = list()

for line in file:
    columns = line.split()
    columns = np.array(columns, dtype=float)
    results.append(columns[:])
    
testing_data = np.array(results)
print(f'\t[INFO] OK!')

print(f'\n[INFO] Getting information about testing dataset...') 
print(f'[INFO] Testing dataset: \n{testing_data}')   
# print(f'[INFO] \tTesting dataset size = {training_data.shape[0]}')
print(f'[INFO] \tTesting dataset size = {testing_data.shape[0]}')


[INFO] Loading testing dataset...
	[INFO] OK!

[INFO] Getting information about testing dataset...
[INFO] Testing dataset: 
[[-0.3565  0.062   5.9891]
 [-0.7842  1.1267  5.5912]
 [ 0.3012  0.5611  5.8234]
 [ 0.7757  1.0648  8.0677]
 [ 0.157   0.8028  6.304 ]
 [-0.7014  1.0316  3.6005]
 [ 0.3748  0.1536  6.1537]
 [-0.692   0.9404  4.4058]
 [-1.397   0.7141  4.9263]
 [-1.8842 -0.2805  1.2548]]
[INFO] 	Testing dataset size = 10


In [99]:
print(f'\n[INFO] Running testing data with Perceptron 1...')
# for inputs in training_data:
for inputs in testing_data:
    result = perceptron1.predict(inputs)
    print(f"[INFO] \tInput: {inputs} -> Output: {result}")


[INFO] Running testing data with Perceptron 1...
[INFO] 	Input: [-0.3565  0.062   5.9891] -> Output: -1
[INFO] 	Input: [-0.7842  1.1267  5.5912] -> Output: 1
[INFO] 	Input: [0.3012 0.5611 5.8234] -> Output: 1
[INFO] 	Input: [0.7757 1.0648 8.0677] -> Output: 1
[INFO] 	Input: [0.157  0.8028 6.304 ] -> Output: 1
[INFO] 	Input: [-0.7014  1.0316  3.6005] -> Output: 1
[INFO] 	Input: [0.3748 0.1536 6.1537] -> Output: -1
[INFO] 	Input: [-0.692   0.9404  4.4058] -> Output: 1
[INFO] 	Input: [-1.397   0.7141  4.9263] -> Output: -1
[INFO] 	Input: [-1.8842 -0.2805  1.2548] -> Output: -1


In [100]:
print(f'\n[INFO] Running testing data with Perceptron 2...')
# for inputs in training_data:
for inputs in testing_data:
    result = perceptron2.predict(inputs)
    print(f"[INFO] \tInput: {inputs} -> Output: {result}")


[INFO] Running testing data with Perceptron 2...
[INFO] 	Input: [-0.3565  0.062   5.9891] -> Output: -1
[INFO] 	Input: [-0.7842  1.1267  5.5912] -> Output: 1
[INFO] 	Input: [0.3012 0.5611 5.8234] -> Output: 1
[INFO] 	Input: [0.7757 1.0648 8.0677] -> Output: 1
[INFO] 	Input: [0.157  0.8028 6.304 ] -> Output: 1
[INFO] 	Input: [-0.7014  1.0316  3.6005] -> Output: 1
[INFO] 	Input: [0.3748 0.1536 6.1537] -> Output: -1
[INFO] 	Input: [-0.692   0.9404  4.4058] -> Output: 1
[INFO] 	Input: [-1.397   0.7141  4.9263] -> Output: -1
[INFO] 	Input: [-1.8842 -0.2805  1.2548] -> Output: -1


Tabela 2 - Resultados com as saídas das saída (classes) para os dados de teste (**tab_teste1.dat**):

<table border="1">
	<tr>
		<td>Amostra</td>
		<td>x1</td>
		<td>x2</td>
		<td>x3</td>
		<td>y (T1)</td>
		<td>y (T2)</td>        
	</tr>
	<tr>
		<td>1</td>
		<td>-0.3565</td>
		<td>0.0620</td>
		<td>5.9891</td>
		<td>-1</td>
		<td>-1</td>
	</tr>
	<tr>
		<td>2</td>
		<td>-0.7842</td>
		<td>1.1267</td>
		<td>5.5912</td>
		<td>1</td>
		<td>1</td>
	</tr>
	<tr>
		<td>3</td>
		<td>0.3012</td>
		<td>0.5611</td>
		<td>5.8234</td>
		<td>1</td>
		<td>1</td>
	</tr>
	<tr>
		<td>4</td>
		<td>0.7757</td>
		<td>1.0648</td>
		<td>8.0677</td>
		<td>1</td>
		<td>1</td>
	</tr>
	<tr>
		<td>5</td>
		<td>0.157</td>
		<td>0.8028</td>
		<td>6.304</td>
		<td>1</td>
		<td>1</td>
	</tr>
	<tr>
		<td>6</td>
		<td>-0.7014</td>
		<td>1.0316</td>
		<td>3.6005</td>
		<td>1</td>
		<td>1</td>
	</tr>            
	<tr>
		<td>7</td>
		<td>0.3748</td>
		<td>0.1536</td>
		<td>6.1537</td>
		<td>-1</td>
		<td>-1</td>
	</tr>
	<tr>
		<td>8</td>
		<td>-0.692</td>
		<td>0.9404</td>
		<td>4.4058</td>
		<td>1</td>
		<td>1</td>
	</tr>
	<tr>
		<td>9</td>
		<td>-1.397</td>
		<td>0.7141</td>
		<td>4.9263</td>
		<td>-1</td>
		<td>-1</td>
	</tr>
	<tr>
		<td>10</td>
		<td>-1.8842</td>
		<td>-0.2805</td>
		<td>1.2548</td>
		<td>-1</td>
		<td>-1</td>
	</tr>                
</table>

*Nota: os valores atuais dessa tabela se referem ao últimos testes realizados. Caso os trechos de código acima sejam executados novamente, necessita-se atualizar a tabela.*

d. Explique por que o número de épocas de treinamento varia a cada vez que se executa o treinamento do perceptron.

*O número de épocas necessárias para a convergência do processo de treinamento varia conforme os valores iniciais dos pesos e do limiar de ativação. Esses valores iniciais definem quão longe da fronteira de seperação o treinamento inicia o ajuste do modelo. Se os valores iniciais atribuídos aos pesos e ao limiar de ativação são definidos aleatoriamente, a cada execução do treinamento, o número de épocas para a convergência normalmente é diferente.*

e. Qual é a principal limitação do perceptron quando aplicado em problemas de classificação de padrões?

*Os perceptrons são um tipo de rede neurais apropriada para tarefa de classificação de padrões em que as classes são linearmente separáveis. Por isso, para problemas não linearmente separáveis utilizando perceptrons, é necessário especificar um número máximo de épocas de treinamento.*

4. Um sistema de gerenciamento automático de controle de duas válvulas, situado a 500 metros de um processo industrial, envia um sinal codificado constituído de quatro grandezas {x1, x2, x3 e x4} que são necessárias para o ajuste de cada uma das válvulas. Conforme mostra a figura abaixo, a mesma via de comunicação é utilizada para acionamento de ambas as válvulas, sendo que o  comutador localizado próximo das válvulas deve decidir se o sinal é para a válvula A ou B. Porém, durante a transmissão, os sinais sofrem interferências que alteram o conteúdo das informações transmitidas. Para resolver este problema, treinar-se-á uma rede ADALINE para classificar os sinais ruidosos, que informará ao sistema comutador se os dados
devem ser encaminhados para o comando de ajuste da válvula A ou B.
Assim, baseado nas medições dos sinais já com ruídos, formou-se o conjunto de treinamento em anexo2, tomando por convenção o valor –1 para os sinais que devem ser encaminhados para o ajuste da válvula A e o valor +1 se os mesmos devem ser enviados para a válvula B.

Daí, pede-se:

a. Execute 2 treinamentos para a rede ADALINE inicializando o vetor de pesos em cada treinamento com valores aleatórios entre zero e um de tal forma que os elementos do vetor de pesos iniciais não sejam os mesmos.

In [3]:
import numpy as np

class Adaline:
    def __init__(self, num_features, learning_rate=0.01, epochs=100, epsilon=1e-5):
        self.num_features = num_features
        self.learning_rate = learning_rate
        self.epochs = epochs
        self.epsilon = epsilon
        self.weights = np.random.rand(num_features + 1)  # initial random weights +1 for the bias term
        print(f'[INFO] \tRandom initial weights: {self.weights}')

    def predict(self, inputs): # activation function
        activation = np.dot(inputs, self.weights[1:]) + self.weights[0] # activation potential: u
        return 1 if activation >= 0 else -1  # Use of the bipolar step function

    def train(self, training_data, targets):
        mse = 0
        for epoch in range(self.epochs):
            total_error = 0
            #print(f'[INFO] Epoch: {epoch}')
            for inputs, target in zip(training_data, targets):
                activation = self.predict(inputs) # return of activation function: y
                error = target - activation
                self.weights[1:] += self.learning_rate * error * inputs
                self.weights[0] += self.learning_rate * error
                total_error += error ** 2
            mse = total_error / len(targets)
            if mse < self.epsilon:
                print(f'[INFO] \tConverged after: {epoch + 1} epochs.')
                break
        print(f'[INFO] \tFinal weights: {self.weights}') 
        print(f'[INFO] \tTotal of epochs: {epoch + 1}')
        print(f'[INFO] \tMean square error: {mse}')

In [4]:
print(f'\n[INFO] ###### Adaline Implementation #######')
print(f'\n[INFO] Loading training dataset and targets...')
file = open('tab_treinamento2.dat', 'r')
results = list()
t = list()

for line in file:
    columns = line.split()
    columns = np.array(columns, dtype=float)
    results.append(columns[:4])
    t.append(columns[-1:])
    
training_data = np.array(results)
targets = np.array(t)
print(f'[INFO] \tOK!')

#training_data = np.array([[4.3290000e-01, -1.3719000e+00, 7.0220000e-01, -8.5350000e-01],[3.0240000e-01, 2.2860000e-01, 8.6300000e-01, 2.7909000e+00],[1.3490000e-01, -6.4450000e-01, 1.0530000e+00, 5.6870000e-01],[3.3740000e-01, -1.7163000e+00, 3.6700000e-01, -6.2830000e-01],[1.1434000e+00, -4.8500000e-02, 6.6370000e-01, 1.2606000e+00],[1.3749000e+00, -5.0710000e-01, 4.4640000e-01, 1.3009000e+00],[7.2210000e-01, -7.5870000e-01, 7.6810000e-01, -5.5920000e-01],[4.4030000e-01, -8.0720000e-01, 5.1540000e-01, -3.1290000e-01],[-5.2310000e-01, 3.5480000e-01, 2.5380000e-01, 1.5776000e+00],[3.2550000e-01, -2.0000000e+00, 7.1120000e-01  -1.1209000e+00],[5.8240000e-01, 1.3915000e+00, -2.2910000e-01, 4.1735000e+00],[1.3400000e-01, 6.0810000e-01, 4.4500000e-01, 3.2230000e+00],[1.4800000e-01, -2.9880000e-01, 4.7780000e-01, 8.6490000e-01],[7.3590000e-01, 1.8690000e-01, -8.7200000e-02, 2.3584000e+00],[7.1150000e-01, -1.1469000e+00, 3.3940000e-01, 9.5730000e-01],[8.2510000e-01, -1.2840000e+00, 8.4520000e-01, 1.2382000e+00],[1.5690000e-01, 3.7120000e-01, 8.8250000e-01, 1.7633000e+00],[3.3000000e-03, 6.8350000e-01, 5.3890000e-01, 2.8249000e+00],[4.2430000e-01, 8.3130000e-01, 2.6340000e-01, 3.5855000e+00],[1.0490000e+00, 1.3260000e-01, 9.1380000e-01, 1.9792000e+00],[1.4276000e+00, 5.3310000e-01, -1.4500000e-02, 3.7286000e+00],[5.9710000e-01, 1.4865000e+00, 2.9040000e-01, 4.6069000e+00],[8.4750000e-01, 2.1479000e+00, 3.1790000e-01, 5.8235000e+00],[1.3967000e+00, -4.1710000e-01, 6.4430000e-01, 1.3927000e+00],[4.4000000e-03, 1.5378000e+00, 6.0990000e-01, 4.7755000e+00],[2.2010000e-01, -5.6680000e-01, 5.1500000e-02, 7.8290000e-01],[6.3000000e-01, -1.2480000e+00, 8.5910000e-01, 8.0930000e-01],[-2.4790000e-01, 8.9600000e-01, 5.4700000e-02, 1.7381000e+00],[-3.0880000e-01, -9.2900000e-02, 8.6590000e-01, 1.5483000e+00],[-5.1800000e-01, 1.4974000e+00, 5.4530000e-01, 2.3993000e+00],[6.8330000e-01, 8.2660000e-01, 8.2900000e-02, 2.8864000e+00],[4.3530000e-01, -1.4066000e+00, 4.2070000e-01, -4.8790000e-01],[-1.0690000e-01, -3.2329000e+00, 1.8560000e-01, -2.4572000e+00],[4.6620000e-01, 6.2610000e-01, 7.3040000e-01, 3.4370000e+00], [8.2980000e-01, -1.4089000e+00, 3.1190000e-01, 1.3235000e+00]])
# targets = np.array([1.0000000e+00, -1.0000000e+00, -1.0000000e+00, -1.0000000e+00, 1.0000000e+00, 1.0000000e+00, 1.0000000e+00, 1.0000000e+00, -1.0000000e+00, 1.0000000e+00, -1.0000000e+00, -1.0000000e+00, 1.0000000e+00, 1.0000000e+00, -1.0000000e+00, -1.0000000e+00, 1.0000000e+00, -1.0000000e+00, -1.0000000e+00, 1.0000000e+00, 1.0000000e+00, -1.0000000e+00, -1.0000000e+00, 1.0000000e+00, -1.0000000e+00, 1.0000000e+00, -1.0000000e+00, 1.0000000e+00, -1.0000000e+00, 1.0000000e+00, 1.0000000e+00, 1.0000000e+00, -1.0000000e+00, -1.0000000e+00, -1.0000000e+00])
print(f'[INFO] \tTraining dataset: \n{training_data}')
print(f'[INFO] \tTargets of training dataset: \n{targets}')


[INFO] ###### Adaline Implementation #######

[INFO] Loading training dataset and targets...
[INFO] 	OK!
[INFO] 	Training dataset: 
[[ 4.3290e-01 -1.3719e+00  7.0220e-01 -8.5350e-01]
 [ 3.0240e-01  2.2860e-01  8.6300e-01  2.7909e+00]
 [ 1.3490e-01 -6.4450e-01  1.0530e+00  5.6870e-01]
 [ 3.3740e-01 -1.7163e+00  3.6700e-01 -6.2830e-01]
 [ 1.1434e+00 -4.8500e-02  6.6370e-01  1.2606e+00]
 [ 1.3749e+00 -5.0710e-01  4.4640e-01  1.3009e+00]
 [ 7.2210e-01 -7.5870e-01  7.6810e-01 -5.5920e-01]
 [ 4.4030e-01 -8.0720e-01  5.1540e-01 -3.1290e-01]
 [-5.2310e-01  3.5480e-01  2.5380e-01  1.5776e+00]
 [ 3.2550e-01 -2.0000e+00  7.1120e-01 -1.1209e+00]
 [ 5.8240e-01  1.3915e+00 -2.2910e-01  4.1735e+00]
 [ 1.3400e-01  6.0810e-01  4.4500e-01  3.2230e+00]
 [ 1.4800e-01 -2.9880e-01  4.7780e-01  8.6490e-01]
 [ 7.3590e-01  1.8690e-01 -8.7200e-02  2.3584e+00]
 [ 7.1150e-01 -1.1469e+00  3.3940e-01  9.5730e-01]
 [ 8.2510e-01 -1.2840e+00  8.4520e-01  1.2382e+00]
 [ 1.5690e-01  3.7120e-01  8.8250e-01  1.7633e+00]


In [5]:
# Creating an Adaline
print(f'\n[INFO] Creating an Adaline...')
number_of_epochs = 10000
epsilon = 1e-5
adaline1 = Adaline(num_features=4, learning_rate=0.01, epochs=number_of_epochs, epsilon=epsilon)
print(f'[INFO] \tOK!')    


[INFO] Creating an Adaline...
[INFO] 	Random initial weights: [0.85687093 0.02633478 0.96817149 0.24457783 0.34848885]
[INFO] 	OK!


In [6]:
# Training the Adaline
print(f'\n[INFO] Getting information about training dataset...')
print(f'[INFO] \tTraining dataset size = {training_data.shape[0]} {training_data.shape[1]}')
print(f'[INFO] \tLabels size = {targets.shape[0]} {targets.shape[1]}')
print(f'[INFO] \tLimit of epochs: {number_of_epochs}')
print(f'[INFO] \tEpsilon: {epsilon}')
print(f'\n[INFO] Training the Adaline 1...')
adaline1.train(training_data, targets)
print(f'[INFO] \tOK!')


[INFO] Getting information about training dataset...
[INFO] 	Training dataset size = 35 4
[INFO] 	Labels size = 35 1
[INFO] 	Limit of epochs: 10000
[INFO] 	Epsilon: 1e-05

[INFO] Training the Adaline 1...
[INFO] 	Converged after: 39 epochs.
[INFO] 	Final weights: [ 0.61687093  0.54973078  0.61733949 -0.08346417 -0.44539315]
[INFO] 	Total of epochs: 39
[INFO] 	Mean square error: [0.]
[INFO] 	OK!


In [7]:
# Creating an Adaline
print(f'\n[INFO] Creating an new Adaline...')
number_of_epochs = 10000
epsilon = 1e-5
adaline2 = Adaline(num_features=4, learning_rate=0.01, epochs=number_of_epochs, epsilon=epsilon)
print(f'[INFO] \tOK!')    


[INFO] Creating an new Adaline...
[INFO] 	Random initial weights: [0.73993445 0.71783308 0.88046206 0.71642348 0.55246439]
[INFO] 	OK!


In [8]:
# Training the Adaline
print(f'\n[INFO] Getting information about training dataset...')
print(f'[INFO] \tTraining dataset size = {training_data.shape[0]} {training_data.shape[1]}')
print(f'[INFO] \tLabels size = {targets.shape[0]} {targets.shape[1]}')
print(f'[INFO] \tLimit of epochs: {number_of_epochs}')
print(f'[INFO] \tEpsilon: {epsilon}')
print(f'\n[INFO] Training the Adaline 2...')
adaline2.train(training_data, targets)
print(f'[INFO] \tOK!')


[INFO] Getting information about training dataset...
[INFO] 	Training dataset size = 35 4
[INFO] 	Labels size = 35 1
[INFO] 	Limit of epochs: 10000
[INFO] 	Epsilon: 1e-05

[INFO] Training the Adaline 2...
[INFO] 	Converged after: 62 epochs.
[INFO] 	Final weights: [ 0.63993445  0.64771308  0.67042606 -0.07245452 -0.49953761]
[INFO] 	Total of epochs: 62
[INFO] 	Mean square error: [0.]
[INFO] 	OK!


b. Registre os resultados dos dois treinamentos na tabela a seguir:

Tabela 3 - Resultados dos treinamentos (**tab_treinamento2.dat**):
<table border="1">
	<tr>
		<td rowspan="2">Treinamento</td>
		<td colspan="5">Vetor de Pesos Inicial</td>
		<td colspan="5">Vetor de Pesos Final</td>
		<td rowspan="2"> Número de Épocas</td>
	</tr>
	<tr>
		<td>b</td>
		<td>w1</td>
		<td>w2</td>
		<td>w3</td>
        <td>w4</td>
		<td>b</td>
		<td>w1</td>
		<td>w2</td>
		<td>w3</td>
		<td>w4</td>        
	</tr>
	<tr>
		<td>1º (T1)</td>
		<td>0.85687093</td>
		<td>0.02633478</td>
		<td>0.96817149</td>
		<td>0.24457783</td>
        <td>0.34848885</td>
		<td>0.61687093</td>
		<td>0.54973078</td>
		<td>0.61733949</td>
		<td>-0.08346417</td>
        <td>-0.44539315</td>
		<td>39</td>
	</tr>
	<tr>
		<td>2º (T1)</td>
		<td>0.73993445</td>
		<td>0.71783308</td>
		<td>0.88046206</td>
		<td>0.71642348</td>
        <td>0.55246439</td>
		<td>0.63993445</td>
		<td>0.64771308</td>
		<td>0.67042606</td>
		<td>-0.07245452</td>
        <td>-0.49953761</td>
		<td>62</td>
	</tr>
</table>

*Nota: os valores atuais dessa tabela se referem ao últimos treinamentos realizados. Caso os trechos de código acima sejam executados novamente, necessita-se atualizar a tabela.*

c. Para os treinamentos realizados, aplique então a rede ADALINE para classificar e informar ao comutador se os sinais seguintes devem ser encaminhados para a válvula A ou B (ver tab_teste2.dat).

In [9]:
# # Testing the adaline
print(f'\n[INFO] Loading testing dataset...')
file = open('tab_teste2.dat', 'r')
results = list()

for line in file:
    columns = line.split()
    columns = np.array(columns, dtype=float)
    results.append(columns[:])
    
testing_data = np.array(results)
print(f'\t[INFO] OK!')

print(f'\n[INFO] Getting information about testing dataset...')
# print(f'[INFO] Testing dataset size = {training_data.shape[0]}')
print(f'[INFO] \tTesting dataset size = {testing_data.shape[0]}')


[INFO] Loading testing dataset...
	[INFO] OK!

[INFO] Getting information about testing dataset...
[INFO] 	Testing dataset size = 15


In [10]:
print(f'\n[INFO] Running testing data with Adaline 1...')
# for inputs in training_data:
for inputs in testing_data:
    result = adaline1.predict(inputs)
    print(f"[INFO] \tInput: {inputs} -> Output: {result}")


[INFO] Running testing data with Adaline 1...
[INFO] 	Input: [0.9694 0.6909 0.4334 3.4965] -> Output: -1
[INFO] 	Input: [0.5427 1.3832 0.639  4.0352] -> Output: -1
[INFO] 	Input: [ 0.6081 -0.9196  0.5925  0.1016] -> Output: 1
[INFO] 	Input: [-0.1618  0.4694  0.203   3.0117] -> Output: -1
[INFO] 	Input: [ 0.187  -0.2578  0.6124  1.7749] -> Output: -1
[INFO] 	Input: [ 0.4891 -0.5276  0.4378  0.6439] -> Output: 1
[INFO] 	Input: [0.3777 2.0149 0.7423 3.3932] -> Output: 1
[INFO] 	Input: [ 1.1498 -0.4067  0.2469  1.5866] -> Output: 1
[INFO] 	Input: [0.9325 1.095  1.0359 3.3591] -> Output: 1
[INFO] 	Input: [0.506  1.3317 0.9222 3.7174] -> Output: -1
[INFO] 	Input: [ 0.0497 -2.0656  0.6124 -0.6585] -> Output: -1
[INFO] 	Input: [0.4004 3.5369 0.9766 5.3532] -> Output: 1
[INFO] 	Input: [-0.1874  1.3343  0.5374  3.2189] -> Output: -1
[INFO] 	Input: [0.506  1.3317 0.9222 3.7174] -> Output: -1
[INFO] 	Input: [ 1.6375 -0.7911  0.7537  0.5515] -> Output: 1


In [11]:
print(f'\n[INFO] Running testing data with Adaline 2...')
# for inputs in training_data:
for inputs in testing_data:
    result = adaline2.predict(inputs)
    print(f"[INFO] \tInput: {inputs} -> Output: {result}")


[INFO] Running testing data with Adaline 2...
[INFO] 	Input: [0.9694 0.6909 0.4334 3.4965] -> Output: -1
[INFO] 	Input: [0.5427 1.3832 0.639  4.0352] -> Output: -1
[INFO] 	Input: [ 0.6081 -0.9196  0.5925  0.1016] -> Output: 1
[INFO] 	Input: [-0.1618  0.4694  0.203   3.0117] -> Output: -1
[INFO] 	Input: [ 0.187  -0.2578  0.6124  1.7749] -> Output: -1
[INFO] 	Input: [ 0.4891 -0.5276  0.4378  0.6439] -> Output: 1
[INFO] 	Input: [0.3777 2.0149 0.7423 3.3932] -> Output: 1
[INFO] 	Input: [ 1.1498 -0.4067  0.2469  1.5866] -> Output: 1
[INFO] 	Input: [0.9325 1.095  1.0359 3.3591] -> Output: 1
[INFO] 	Input: [0.506  1.3317 0.9222 3.7174] -> Output: -1
[INFO] 	Input: [ 0.0497 -2.0656  0.6124 -0.6585] -> Output: -1
[INFO] 	Input: [0.4004 3.5369 0.9766 5.3532] -> Output: 1
[INFO] 	Input: [-0.1874  1.3343  0.5374  3.2189] -> Output: -1
[INFO] 	Input: [0.506  1.3317 0.9222 3.7174] -> Output: -1
[INFO] 	Input: [ 1.6375 -0.7911  0.7537  0.5515] -> Output: 1


Tabela 4 - Resultados com as saídas das saída (classes) para os dados de teste (**tab_teste2.dat**):

<table border="1">
	<tr>
		<td>Amostra</td>
		<td>x1</td>
		<td>x2</td>
		<td>x3</td>
        <td>x4</td>
		<td>y (T1)</td>
		<td>y (T2)</td>        
	</tr>
	<tr>
		<td>1</td>
		<td>0.9694</td>
		<td>0.6909</td>
		<td>0.4334</td>
        <td>3.4965</td>
		<td>-1</td>
		<td>-1</td>
	</tr>
	<tr>
		<td>2</td>
		<td>0.5427</td>
		<td>1.3832</td>
		<td>0.639</td>
		<td>4.0352</td>
		<td>-1</td>
        <td>-1</td>
	</tr>
	<tr>
		<td>3</td>
		<td>0.6081</td>
		<td>-0.9196</td>
		<td>0.5925</td>
		<td>0.1016</td>
		<td>1</td>
        <td>1</td>
    </tr>
	<tr>
		<td>4</td>
		<td>-0.1618</td>
		<td>0.4694</td>
		<td>0.203</td>
		<td>3.0117</td>
		<td>-1</td>
        <td>-1</td>
    </tr>
	<tr>
		<td>5</td>
		<td>0.187</td>
		<td>-0.2578</td>
		<td>0.6124</td>
		<td>1.7749</td>
		<td>-1</td>
        <td>-1</td>
    </tr>
	<tr>
		<td>6</td>
		<td>0.4891</td>
		<td>-0.5276</td>
		<td>0.4378</td>
		<td>0.6439</td>
		<td>1</td>
        <td>1</td>
    </tr>
	<tr>
		<td>7</td>
		<td>0.3777</td>
		<td>2.0149</td>
		<td>0.7423</td>
		<td>3.3932</td>
		<td>1</td>
        <td>1</td>
    </tr>
	<tr>
		<td>8</td>
		<td>1.1498</td>
		<td>-0.4067</td>
		<td>0.2469</td>
		<td>1.5866</td>
		<td>1</td>
        <td>1</td>
    </tr>
	<tr>
		<td>9</td>
		<td>0.9325</td>
		<td>1.095</td>
		<td>1.0359</td>
		<td>3.3591</td>
		<td>1</td>
        <td>1</td>
    </tr>
	<tr>
		<td>10</td>
		<td>0.506</td>
		<td>1.3317</td>
		<td>0.9222</td>
		<td>3.7174</td>
		<td>-1</td>
        <td>-1</td>
    </tr>
	<tr>
		<td>11</td>
		<td>0.0497</td>
		<td>-2.0656</td>
		<td>0.6124</td>
		<td>-0.6585</td>
		<td>-1</td>
        <td>-1</td>
    </tr> 
	<tr>
		<td>12</td>
		<td>0.4004</td>
		<td>3.5369</td>
		<td>0.9766</td>
		<td>5.3532</td>
		<td>1</td>
        <td>1</td>
    </tr>
	<tr>
		<td>13</td>
		<td>-0.1874</td>
		<td>1.3343</td>
		<td>0.5374</td>
		<td>3.2189</td>
		<td>-1</td>
        <td>-1</td>
    </tr>
	<tr>
		<td>14</td>
		<td>0.506</td>
		<td>1.3317</td>
		<td>0.9222</td>
		<td>3.7174</td>
		<td>-1</td>
        <td>-1</td>
    </tr>
	<tr>
		<td>15</td>
		<td>1.6375</td>
		<td>-0.7911</td>
		<td>0.7537</td>
		<td>0.5515</td>
		<td>1</td>
        <td>1</td>
    </tr> 					
</table>

*Nota: os valores atuais dessa tabela se referem ao últimos testes realizados. Caso os trechos de código acima sejam executados novamente, necessita-se atualizar a tabela.*

5. Um(a) estudante da disciplina de Redes Neurais e Aprendizado Profundo ficou empolgado(a) com o trabalho do Fisher sobre as flores Íris e resolveu propor uma versão automatizada para ele. Essa nova versão deveria ter dois módulos principais: um módulo de visão computacional e um módulo do tipo classificador neural. Caso você(s) fosse(m) esse(a) estudante, como você(s) desenvolveria(m) esse sistema? Descreva-o em detalhes. Use ilustração(ões) para valorizar o seu pré-projeto. Lembre-se que são três tipos de Íris (Virginica, Versicolor e Setosa) e que 4 parâmetros foram medidos pelo Fisher para cada uma das flores (comprimento e largura da Pétala, Comprimento e largura da Sépala).

6. Considere a base de dados encontrada em Irisdat.xlsx. Daí, pede-se: a) Treinar um PMC que classifique observações de flores íris em 3 espécies (Setosa, Versicolor e Virginica) usando como entradas as características SEPALLENGTH (SL), SEPALWIDTH (SW), PETALLENGTH (PL) e PETALWIDTH (PW). b) Estime SL a partir de SW, PL, PW.

7. Considere a base de dados encontrada em engines.xlsx, em que ‘Fuel rate’ e ‘Speed’ são variáveis de entrada e ‘Torque’ e ‘Nitrous Oxide Emissions (NOE)’ são as variáveis de saída, respectivamente. Desenvolva três regressores. Um deles deve estimar conjuntamente o ‘Torque’ e o NOE. Já os outros dois devem estimar essas saídas separadamente (i.e. um estimará o Torque e o outro o NOE). Compare o desempenho das duas estratégias apontando qual delas apresenta uma maior capacidade de generalização.

8. Valendo-se da base de dados reais referente ao Volume de Vendas de Passagens (VVP) de uma companhia aérea norte-americana que se encontra no arquivo vvp.xlsx, pede-se: 1) Desenvolver um previsor neural que receba como entradas os VVPs registrados nos instantes k-1 e k-12 (i.e. VVP(k-1) e VVP(k-12)) e que disponibilize na saída o VVP no instante corrente k (i.e. VVP(k)). O previsor deverá realizar previsões recursivas de 1 a 12 passos à frente (i.e., de um a doze meses à frente); 2) De posse da base de dados, remova a tendência linear presente na base de dados original. Desse modo, você conhecerá a série destendenciada e a tendência linear. Para a primeira série, desenvolva um previsor neural que receba como entradas os VVPs registrados nos instantes k-1 e k-12 (i.e. VVP(k-1) e VVP(k-12)) e que disponibilize na saída o VVP no instante corrente k (i.e. VVP(k)). O previsor deverá realizar previsões recursivas de 1 a 12 passos à frente (i.e., de um a doze meses à frente). Para a segunda (i.e., a tendência linear), preveja linearmente os próximos dozes pontos. Em seguida, some ponto a ponto as duas previsões e compare o desempenho dessa abordagem com a anterior apontando qual delas apresenta uma maior capacidade de generalização.

9. Procure na literatura 2 artigos que tratem do tema Sensores Inferenciais (ou Soft Sensors) para uma dada grandeza de seu interesse (e.g. temperatura, pressão, vazão, nível etc.) e que tenham sido publicados nos últimos 5 anos. Explique de forma sucinta o que foi desenvolvido pelos autores, referenciando-os. Sugestão: As principais informações de qualquer artigo geralmente se encontram no título, no resumo e nas conclusões. Ao ler esses três itens, o leitor tem uma boa ideia do que esperar daquele trabalho. A propósito, usualmente o leitor decidirá se lerá todo o artigo ou não com base na sua impressão a respeito desses três itens.